### PR TITLE
Tests: Contract testing for the procedure AddTimeData and related fixes

### PR DESCRIPTION
**Pull request by mmachicao**
Objective: Make the contract for AddTimeData testable.

The procedure AddTimeData is stateful, so any testing beyond trivialities requires some modification in the code:

- Separation between orchestration (lock/unlock) and algorithm
- Pushing static (stateful) variables out of the procedure body
- Applied clang-format-diff.py

Additionally:
wallet_tests.cpp contains errors -> Tests for ComputeTimeSmart falsely assumed that time offset is zero

### CODE REVIEW

File: src/test/timedata_tests.cpp [hunk 1]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+    BOOST_CHECK(neq);
+
+    neq = !(cn2 == cn3);
+
+    BOOST_CHECK(neq);
+}
+
+BOOST_AUTO_TEST_CASE(util_AddTimeDataAlgorithmComputeOffsetWhenNewSampleCountIsGreaterEqualFiveAndUneven)
+{
+    int capacity = 20; //  can store up to 20 samples
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Should be unsigned to match what `CMedianFilter` expects?

Applies for all `int capacity` in this file.

File: src/test/timedata_tests.cpp [hunk 2]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+    int64_t offset = 0;
+    std::set<CNetAddr> knownSet;
+    CMedianFilter<int64_t> offsetFilter(capacity, 0); // max size : 20 , init sample: 0
+
+
+    for (unsigned int sample = 1; sample < 10; sample++) { // precondition: at least 4 samples. (including the init sample : 0)
+        CNetAddr addr = UtilBuildAddress(0x001, 0x001, 0x001, sample);
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Make the implicit conversion from `unsigned int` to `unsigned char` explicit here.

Applies for all `CNetAddr addr = UtilBuildAddress(…, sample);` in this file.

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -290,25 +290,52 @@ static int64_t AddTx(CWallet& wallet, uint32_t lockTime, int64_t mockTime, int64
 // expanded to cover more corner cases of smart time logic.
 BOOST_AUTO_TEST_CASE(ComputeTimeSmart)
 {
+    // The ComputeTimeSmart function within the wallet reads the current time using the GetAdjustedTime function in timedata.
+    //
+    // This function computes the time based on the system time (which suports mocking)
```
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: Should be “supports” :-)

File: src/wallet/test/wallet_tests.cpp [hunk 4]
```diff
@@ -290,25 +290,52 @@ static int64_t AddTx(CWallet& wallet, uint32_t lockTime, int64_t mockTime, int64
...
+    int64_t clockTime = GetAdjustedTime(); // time + time offset (unfortunately, from a statefull data structure)
+    int64_t blockTime = clockTime + 5;
+
+    BOOST_CHECK_EQUAL(AddTx(m_wallet, 1, blockTime), clockTime); // clocktime shall be used
 
     // Test that updating existing transaction does not change smart time.
-    BOOST_CHECK_EQUAL(AddTx(m_wallet, 1, 200, 220), 100);
+    SetMockTime(20);
+    int64_t newBlockTime = GetAdjustedTime() + 10;
+    BOOST_CHECK_EQUAL(AddTx(m_wallet, 1, newBlockTime), clockTime); // time has not changed from the preceeding transaction
```
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: Should be “preceding” :-)

File: src/wallet/test/wallet_tests.cpp [hunk 5]
```diff
@@ -290,25 +290,52 @@ static int64_t AddTx(CWallet& wallet, uint32_t lockTime, int64_t mockTime, int64
...
+    clockTime = GetAdjustedTime();
+    blockTime = clockTime - 2;
+    BOOST_CHECK_EQUAL(AddTx(m_wallet, 3, blockTime), blockTime); // blocktime shall be used
 
     // New transaction should use latest entry time if higher than
     // min(block time, clock time).
-    BOOST_CHECK_EQUAL(AddTx(m_wallet, 4, 500, 390), 400);
+    SetMockTime(5);
+    newBlockTime = blockTime - 5;
+    BOOST_CHECK_EQUAL(AddTx(m_wallet, 4, newBlockTime), blockTime); // using the blocktime of the preceeding transaction
```
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: Should be “preceding” :-)

File: src/timedata.cpp [hunk 6]
```diff
@@ -44,20 +44,36 @@ static int64_t abs64(int64_t n)
 
 #define BITCOIN_TIMEDATA_MAX_SAMPLES 200
 
+
+static std::set<CNetAddr> setKnown;
+
+static CMedianFilter<int64_t> vTimeOffsets(BITCOIN_TIMEDATA_MAX_SAMPLES, 0);
+
+int CountOffsetSamples()
```
Comment (practicalswift) [inline src/timedata.cpp]: Unused. Drop this :-)

File: src/timedata.h [hunk 7]
```diff
@@ -75,4 +76,8 @@ int64_t GetTimeOffset();
 int64_t GetAdjustedTime();
 void AddTimeData(const CNetAddr& ip, int64_t nTime);
 
+/** Functions to isolate the contract in AddTimeData */
+int CountOffsetSamples();
```
Comment (practicalswift) [inline src/timedata.h]: Unused :-)

File: src/wallet/test/wallet_tests.cpp [hunk 8]
```diff
@@ -290,25 +290,52 @@ static int64_t AddTx(CWallet& wallet, uint32_t lockTime, int64_t mockTime, int64
...
+    // This function computes the time based on the system time (which suports mocking)
+    // and the time offset, which is not necessarily zero (and does not support mocking)
+    //
+    // Tests have to be executed in order, because the individual cases are stateful
+
+
     // New transaction should use clock time if lower than block time.
-    BOOST_CHECK_EQUAL(AddTx(m_wallet, 1, 100, 120), 100);
+    SetMockTime(10);
+    int64_t clockTime = GetAdjustedTime(); // time + time offset (unfortunately, from a statefull data structure)
```
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: Should be “stateful” :-)

File: src/test/timedata_tests.cpp [hunk 9]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+    BOOST_CHECK_EQUAL(offsetFilter.size(), samples + 1); // sample was added...
+    BOOST_CHECK(oldOffset == offset);                    // ...but new offset was not computed
+}
+
+BOOST_AUTO_TEST_CASE(util_AddTimeDataAlgorithmComputeOffsetWithMedianWithinBounds)
+{
+    int capacity = 10;
+    int64_t offset = 0;
+    std::set<CNetAddr> knownSet;
+    CMedianFilter<int64_t> offsetFilter(capacity, 0); // max size : 10 , init samplee: 0
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Should be “sample”?

File: src/test/timedata_tests.cpp [hunk 10]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+
+
+    for (unsigned int sample = 1; sample < 10; sample++) { // precondition: at least 4 samples. (including the init sample : 0)
+        CNetAddr addr = UtilBuildAddress(0x001, 0x001, 0x001, sample);
+        AddTimeDataAlgorithm(addr, sample, knownSet, offsetFilter, offset); // 1.1.1.[1,2,3],  offsetSample = [1,2,3]
+    }
+
+    BOOST_CHECK(offset != 0); // offset has changed from the initial value
+
+    assert(offsetFilter.size() == 10); // next sample will be the 11th (uneven) and will trigger a new computation of the offset
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Why `assert(…)` and not `BOOST_CHECK(…)`? Applies throughout this PR.

File: src/test/timedata_tests.cpp [hunk 11]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+    for (unsigned int sample = 1; sample < 9; sample++) { // precondition: at least 7 samples. (including the init sample : 0
+        CNetAddr addr = UtilBuildAddress(0x001, 0x001, 0x001, sample);
+        AddTimeDataAlgorithm(addr, sample, knownSet, offsetFilter, offset); // 1.1.1.[1,2,3],  offsetSample = [1,2,3]
+    }
+
+    BOOST_CHECK(offset != 0); // offset has changed from the initial value
+
+    assert(offsetFilter.size() == 9); // next sample will be the 10th (even) and will *not* trigger a new computation of the offset
+
+
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Drop one of the newlines. Applies for all double newlines in this PR.

File: src/test/timedata_tests.cpp [hunk 12]
```diff
@@ -34,4 +35,223 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
...
+{
+    int capacity = 10;
+    int64_t offset = 0;
+    std::set<CNetAddr> knownSet;
+    CMedianFilter<int64_t> offsetFilter(capacity, 0); // max size : 10 , init sample: 0
+
+    for (int sample = 1; sample < 4; sample++) {                                                     // precondition: 4 samples, all outside bounds
+        CNetAddr addr = UtilBuildAddress(0x001, 0x001, 0x001, sample);                               // 1.1.1.[1,2,3]
+        AddTimeDataAlgorithm(addr, 2 * DEFAULT_MAX_TIME_ADJUSTMENT, knownSet, offsetFilter, offset); // offsetSample  = 2 * DEFAULT_MAX_TIME_ADJUSTMENT (out of bounds)
+
```
Comment (practicalswift) [inline src/test/timedata_tests.cpp]: Drop empty line at end of block.

**Comment by fanquake** [comment]
Going to close as "Up for Grabs". @dongcarl You might be interested in having a look here.

**Comment by adamjonas** [comment]
Believe https://github.com/bitcoin/bitcoin/pull/16563 removes the need for this. If so, "up for grabs" label should be removed.

**Pull request closed without merging**